From 4592cd8e763e94173a5389461b3895bcbc64a305 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Fri, 24 Apr 2026 09:43:19 -0300
Subject: [PATCH] Add properties to disable some features/commands/states that
 the ble chip declared but doesnt actually support

Ported to android-16.0.0_r4 by bariccattion.
---
 system/gd/hci/controller_impl.cc | 67 ++++++++++++++++++++++++++++++--
 1 file changed, 64 insertions(+), 3 deletions(-)

diff --git a/system/gd/hci/controller_impl.cc b/system/gd/hci/controller_impl.cc
index 69b2294bbe..87a16fce89 100644
--- a/system/gd/hci/controller_impl.cc
+++ b/system/gd/hci/controller_impl.cc
@@ -35,6 +35,9 @@
 #include "sysprops/sysprops_module.h"
 #endif
 
+#include <sstream>
+using bluetooth::os::GetSystemProperty;
+
 namespace bluetooth {
 namespace hci {
 
@@ -365,6 +367,19 @@ struct ControllerImpl::impl {
     ErrorCode status = complete_view.GetStatus();
     log::assert_that(status == ErrorCode::SUCCESS, "Status {}", ErrorCodeText(status));
     local_supported_commands_ = complete_view.GetSupportedCommands();
+    std::string ignored_commands = GetSystemProperty("persist.sys.bt.unsupported.commands").value_or("");
+
+    if (ignored_commands != "") {
+        std::stringstream ss(ignored_commands);
+        std::string command;
+        while (std::getline(ss, command, ',')) {
+            int index = std::stoi(command);
+            log::warn("Ignoring local supported command {}", index);
+            uint16_t byte_index = index / 10;
+            uint16_t bit_index = index % 10;
+            local_supported_commands_[byte_index] &= ~(1 << bit_index);
+        }
+    }
   }
 
   void read_local_extended_features_complete_handler(std::promise<void> promise,
@@ -374,7 +390,21 @@ struct ControllerImpl::impl {
     ErrorCode status = complete_view.GetStatus();
     log::assert_that(status == ErrorCode::SUCCESS, "Status {}", ErrorCodeText(status));
     uint8_t page_number = complete_view.GetPageNumber();
-    extended_lmp_features_array_.push_back(complete_view.GetExtendedLmpFeatures());
+    auto lmp_features = complete_view.GetExtendedLmpFeatures();
+    std::string ignored_features = GetSystemProperty("persist.sys.bt.unsupported.ogfeatures").value_or("");
+
+    if (ignored_features != "") {
+        std::stringstream ss(ignored_features);
+        std::string feature;
+        while (std::getline(ss, feature, ',')) {
+            int index = std::stoi(feature) - (page_number * 64);
+            if(index >= 0 && index < 64) {
+                log::warn("Ignoring local supported feature {}", index);
+                lmp_features &= ~(1ULL << index);
+            }
+        }
+    }
+    extended_lmp_features_array_.push_back(lmp_features);
     if (page_number == 0 && local_version_information_.manufacturer_name_ == LMP_COMPID_INTEL &&
         local_version_information_.lmp_version_ == LmpVersion::V_4_2 &&
         local_version_information_.lmp_subversion_ == LMP_SUBVERSION_INTEL_AC7265) {
@@ -537,7 +567,19 @@ struct ControllerImpl::impl {
     ASSERT(complete_view.IsValid());
     ErrorCode status = complete_view.GetStatus();
     log::assert_that(status == ErrorCode::SUCCESS, "Status {}", status, ErrorCodeText(status));
-    le_local_supported_features_ = complete_view.GetLeFeatures();
+    auto local_features = complete_view.GetLeFeatures();
+    std::string ignored_features = GetSystemProperty("persist.sys.bt.unsupported.lefeatures").value_or("");
+
+    if (ignored_features != "") {
+        std::stringstream ss(ignored_features);
+        std::string feature;
+        while (std::getline(ss, feature, ',')) {
+            int index = std::stoi(feature);
+            log::warn("Ignoring local supported feature {}", index);
+            local_features &= ~(1ULL << index);
+        }
+    }
+    le_local_supported_features_ = local_features;
   }
 
   void le_read_supported_states_handler(CommandCompleteView view) {
@@ -545,7 +588,19 @@ struct ControllerImpl::impl {
     ASSERT(complete_view.IsValid());
     ErrorCode status = complete_view.GetStatus();
     log::assert_that(status == ErrorCode::SUCCESS, "Status {}", ErrorCodeText(status));
-    le_supported_states_ = complete_view.GetLeStates();
+    auto local_states = complete_view.GetLeStates();
+    std::string ignored_states = GetSystemProperty("persist.sys.bt.unsupported.states").value_or("");
+
+    if (ignored_states != "") {
+        std::stringstream ss(ignored_states);
+        std::string state;
+        while (std::getline(ss, state, ',')) {
+            int index = std::stoi(state);
+            log::warn("Ignoring local supported state {}", index);
+            local_states &= ~(1ULL << index);
+        }
+    }
+    le_supported_states_ = local_states;
   }
 
   void le_read_accept_list_size_handler(CommandCompleteView view) {
-- 
2.53.0.windows.3
